File: src/space/cherryband/ari/LookupFragment.java
package space.cherryband.ari;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import androidx.core.view.MenuItemCompat;
import androidx.appcompat.widget.SearchView;
import androidx.fragment.app.FragmentActivity;

import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import org.jetbrains.annotations.NotNull;

import java.util.Timer;
import java.util.TimerTask;

import itkach.slob.Slob;

public class LookupFragment extends BaseListFragment implements LookupListener {

    private Timer timer;
    private SearchView searchView;
    private AriApplication app;
    private SearchView.OnQueryTextListener queryTextListener;
    private SearchView.OnCloseListener closeListener;
    private MenuItemCompat.OnActionExpandListener openListener;

    private final static String TAG = LookupFragment.class.getSimpleName();

    @Override
    char getEmptyIcon() {
        return IconMaker.IC_SEARCH;
    }

    @Override
    CharSequence getEmptyText() {
        return "";
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        app = (AriApplication) getActivity().getApplication();
        app.addLookupListener(this);
    }

    @Override
    protected boolean supportsSelection() {
        return false;
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setBusy(false);
        ListView listView = getListView();
        listView.setOnItemClickListener((parent, view1, position, id) -> {
            Log.i("--", "Item clicked: " + position);
            Intent intent = new Intent(getActivity(),
                    ArticleCollectionActivity.class);
            intent.putExtra("position", position);
            startActivity(intent);
        });
        final AriApplication app = (AriApplication) getActivity().getApplication();
        getListView().setAdapter(app.lastResult);

        closeListener = () -> true;

        queryTextListener = new SearchView.OnQueryTextListener() {

            TimerTask scheduledLookup = null;

            @Override
            public boolean onQueryTextSubmit(String query) {
                Log.d(TAG, "query text submit: " + query);
                onQueryTextChange(query);
                return true;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                Log.d(TAG, "new query text: " + newText);
                TimerTask doLookup = new TimerTask() {
                    @Override
                    public void run() {
                        final String query = searchView.getQuery().toString();
                        if (app.getLookupQuery().equals(query)) {
                            return;
                        }
                        getActivity().runOnUiThread(() -> app.lookup(query));
                        scheduledLookup = null;
                    }
                };
                final String query = searchView.getQuery().toString();
                if (!app.getLookupQuery().equals(query)) {
                    if (scheduledLookup != null) {
                        scheduledLookup.cancel();
                    }
                    scheduledLookup = doLookup;
                    timer.schedule(doLookup, 600);
                }
                return true;
            }
        };

    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.action_random){
            AriApplication app = (AriApplication) getActivity().getApplication();
            Slob.Blob blob = app.random();
            if (blob == null) {
                Toast.makeText(getContext(),
                        R.string.article_collection_nothing_found,
                        Toast.LENGTH_SHORT).show();
                return true;
            }
            Intent intent = new Intent(getActivity(), ArticleCollectionActivity.class);
            intent.setData(Uri.parse(app.getUrl(blob)));
            startActivity(intent);
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        timer = new Timer();
        inflater.inflate(R.menu.lookup, menu);
        MenuItem miFilter = menu.findItem(R.id.action_lookup);
        View filterActionView = miFilter.getActionView();
        searchView = (SearchView) filterActionView.findViewById(R.id.fldLookup);
        searchView.setQueryHint(miFilter.getTitle());
        searchView.setOnQueryTextListener(queryTextListener);
        searchView.setOnCloseListener(closeListener);
        searchView.setSubmitButtonEnabled(false);
    }

    @Override
    public void onPrepareOptionsMenu(Menu menu) {
        FragmentActivity activity = getActivity();
        super.onPrepareOptionsMenu(menu);
        if (app.getAutoPaste()) {
            CharSequence clipboard = Clipboard.take(activity);
            if (clipboard != null) {
                app.lookup(clipboard.toString(), false);
            }
        }
        CharSequence query = app.getLookupQuery();
        searchView.setQuery(query, true);
        if (app.lastResult.getCount() > 0) {
            searchView.clearFocus();
        }
        MenuItem miRandomArticle = menu.findItem(R.id.action_random);
        miRandomArticle.setIcon(IconMaker.actionBar(activity, IconMaker.IC_RELOAD));
    }

    @Override
    public void onSaveInstanceState(@NotNull Bundle outState) {
        super.onSaveInstanceState(outState);
        if (searchView != null) {
            String query = searchView.getQuery().toString();
            outState.putString("lookupQuery", query);
        }
    }

    private void setBusy(boolean busy) {
        setListShown(!busy);
        if (!busy) {
            TextView emptyText = emptyView.findViewById(R.id.empty_text);
            String msg = "";
            String query = app.getLookupQuery();
            if (query != null && !query.equals("")) {
                msg = getString(R.string.lookup_nothing_found);
            }
            emptyText.setText(msg);
        }
    }

    @Override
    public void onDestroy() {
        if (timer != null) {
            timer.cancel();
        }
        AriApplication app = (AriApplication) getActivity().getApplication();
        app.removeLookupListener(this);
        super.onDestroy();
    }

    @Override
    public void onLookupStarted(String query) {
        setBusy(true);
    }

    @Override
    public void onLookupFinished(String query) {
        setBusy(false);
    }

    @Override
    public void onLookupCanceled(String query) {
        setBusy(false);
    }

}

File: src/space/cherryband/ari/LookupFragment.kt
package space.cherryband.ari

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.AdapterView.OnItemClickListener
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.widget.SearchView
import java.util.*

class LookupFragment : BaseListFragment(), LookupListener {
    private var timer: Timer? = null
    private lateinit var searchView: SearchView

    override val emptyIcon: Char = IconMaker.IC_SEARCH
    override val emptyText = ""
    override val supportsSelection = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val app = activity!!.application as AriApplication
        app.addLookupListener(this)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setBusy(false)
        val listView = listView
        listView.onItemClickListener =
            OnItemClickListener { _: AdapterView<*>?, _: View?, position: Int, _: Long ->
                Log.i("--", "Item clicked: $position")
                Intent(activity, ArticleCollectionActivity::class.java)
                    .putExtra("position", position)
                    .let { startActivity(it) }
            }
        val app = activity!!.application as AriApplication
        getListView().adapter = app.lastResult
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.action_random) {
            val app = activity!!.application as AriApplication
            val blob = app.random()
            if (blob == null) {
                Toast.makeText(
                    context,
                    R.string.article_collection_nothing_found,
                    Toast.LENGTH_SHORT
                ).show()
                return true
            }
            val intent = Intent(activity, ArticleCollectionActivity::class.java)
            intent.data = Uri.parse(app.getUrl(blob))
            startActivity(intent)
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private val closeListener = SearchView.OnCloseListener { true }
    private val queryTextListener = object : SearchView.OnQueryTextListener {
        var scheduledLookup: TimerTask? = null
        override fun onQueryTextSubmit(query: String): Boolean {
            Log.d(TAG, "query text submit: $query")
            onQueryTextChange(query)
            return true
        }

        override fun onQueryTextChange(newText: String): Boolean {
            Log.d(TAG, "new query text: $newText")
            val app = activity!!.application as AriApplication
            val doLookup: TimerTask = object : TimerTask() {
                override fun run() {
                    val query = searchView.query.toString()
                    if (app.lookupQuery == query) { return }
                    activity!!.runOnUiThread { app.lookup(query) }
                    scheduledLookup = null
                }
            }
            val query = searchView.query.toString()
            if (app.lookupQuery != query) {
                scheduledLookup?.cancel()
                scheduledLookup = doLookup
                timer!!.schedule(doLookup, 600)
            }
            return true
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        timer = Timer()
        inflater.inflate(R.menu.lookup, menu)
        val miFilter = menu.findItem(R.id.action_lookup)
        val filterActionView = miFilter.actionView
        searchView = filterActionView.findViewById(R.id.fldLookup)
        searchView.apply {
            queryHint = miFilter.title
            setOnQueryTextListener(queryTextListener)
            setOnCloseListener(closeListener)
            isSubmitButtonEnabled = false
        }
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        val activity = activity
        super.onPrepareOptionsMenu(menu)
        val app = activity!!.application as AriApplication
        if (app.autoPaste) {
            val clipboard = Clipboard.take(activity)
            if (clipboard != null) {
                app.lookup(clipboard.toString(), false)
            }
        }
        val query: CharSequence = app.lookupQuery
        searchView.setQuery(query, true)
        if (app.lastResult!!.count > 0) {
            searchView.clearFocus()
        }
        val miRandomArticle = menu.findItem(R.id.action_random)
        miRandomArticle.icon = IconMaker.actionBar(activity, IconMaker.IC_RELOAD)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        val query = searchView.query.toString()
        outState.putString("lookupQuery", query)
    }

    private fun setBusy(busy: Boolean) {
        setListShown(!busy)
        if (!busy) {
            val app = activity!!.application as AriApplication
            val emptyText = emptyView.findViewById<TextView>(R.id.empty_text)
            var msg = ""
            val query = app.lookupQuery
            if (query != "") {
                msg = getString(R.string.lookup_nothing_found)
            }
            emptyText.text = msg
        }
    }

    override fun onDestroy() {
        if (timer != null) {
            timer!!.cancel()
        }
        val app = activity!!.application as AriApplication
        app.removeLookupListener(this)
        super.onDestroy()
    }

    override fun onLookupStarted(query: String) = setBusy(true)
    override fun onLookupFinished(query: String) = setBusy(false)
    override fun onLookupCanceled(query: String) = setBusy(false)

    companion object {
        private val TAG = LookupFragment::class.java.simpleName
    }
}